### PR TITLE
Added tests where multiple GC runs happen for a single successful summarization attempt due to retries

### PR DESCRIPTION
**Pull request by agarwal-navin**
#18632 added GC sweep ops. As called out in the Todo section of that PR, if GC succeeds but a follow up step fails during summarization, the next GC run may generate an sweep op that contains the same deleted nodes id. This PR adds tests to validate these two scenarios:
1. GC succeeds but summary fails multiple times on summarization retries until the final summary attempt which is successful. A sweep op is sent during the first attempt which may get acked during the attempts. When it does, not further sweep ops are sent.
2. GC succeeds but summary fails multiple times on summarization retries until the final summary attempt which is successful. The sweep ops are not processed (inbound queue is paused) so every GC attempt ends up sending a sweep op with the same deleted node ids.

Also, update `summarizeOnDemand` to support running with retries in case of failures.

[AB#6547](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/6547)

### CODE REVIEW

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 1]
```diff
@@ -770,10 +770,16 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
 		const maybeUnusedStorageIds: Set<string> = new Set();
 		for (const route of blobRoutes) {
 			const blobId = getBlobIdFromGCNodePath(route);
+			// If the blob hasn't already been deleted, log an error because this should never happen.
+			// If the blob has already been deleted, log a telemetry event. This can happen because multiple GC
+			// sweep ops can contain the same data store. It would be interesting to track how often this happens.
+			const alreadyDeleted = this.isBlobDeleted(route);
 			if (!this.redirectTable.has(blobId)) {
-				this.mc.logger.sendErrorEvent({
+				this.mc.logger.sendTelemetryEvent({
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/blobManager.ts:778]: Why is this now just a Generic instead of an Error Event?

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts [hunk 2]
```diff
@@ -29,48 +31,110 @@ import {
...
+	);
+}
 
 /**
  * These tests validate that SweepReady data stores are correctly swept. Swept datastores should be
  * removed from the summary, added to the GC deleted blob, and prevented from changing (sending / receiving ops,
  * loading, etc.).
  */
 describeCompat("GC data store sweep tests", "NoCompat", (getTestObjectProvider) => {
-	const remainingTimeUntilSweepMs = 100;
```
Comment (tyler-cai-microsoft) [inline packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:41]: It looks like we got rid of the tests that were single container loading right before any timeouts. Do we still have coverage somewhere?

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts [hunk 3]
```diff
@@ -541,23 +541,245 @@ describeCompat("GC data store sweep tests", "NoCompat", (getTestObjectProvider)
...
 			// summarize and GC succeed.
 			await assert.doesNotReject(
-				async () => summarizeNow(summarizer),
+				async () => summarize(summarizer),
 				"Summary and GC should succeed with deleted data store",
 			);
 		});
 	});
+
+	describe("Sweep with summarize failures and retries", () => {
```
Comment (tyler-cai-microsoft) [inline packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:562]: Since the scenario we're trying to resolve is creating a datastore on the first summarize run, could we add a test to do that here so that we can see if we fix that scenario?

File: packages/runtime/container-runtime/src/blobManager.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/blobManager.ts:778]: The comment above explains. I added the "category" property explicitly. It's generic only if the blob has already been deleted. This can happen as there can be multiple sweep ops witht the same deleted node id.

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts [hunk 4]
```diff
@@ -29,48 +31,110 @@ import {
...
+
+/**
+ * Validates that the given data store state is correct in the summary.
+ * If expectDelete is true:
+ * - It should be deleted from the data store summary tree.
+ * If expectGCStateHandle is not true:
+ * - It should not be present in the GC state in GC summary tree.
+ * - It should be present in the deleted nodes in GC summary tree.
+ */
+function validateDataStoreStateInSummary(
```
[the remark this answers is not in the document]
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:55]: I feel the same way :-). We have a plan to go through all the GC tests and update them since the validations have evolved over time. We should make them all use similar validations and share logic as much as possible.

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:41]: I didn't see those tests in this file. I assumed that this part I removed was carry over from other tests where it is actually used.

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts (hunk 3 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:562]: That's not what these tests are for. They are simply testing that if multiple GC ops are sent with the same deleted node ids, things still work as expected. It doesn't care about what the failure is that leads to it.

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts (hunk 2 above)
Comment (tyler-cai-microsoft) [reply packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:41]: Ok, as long as we have coverage, I'm good

**Review by tyler-cai-microsoft — APPROVED** [review]

**Comment by markfields** [comment]
@agarwal-navin - Looking at the description, test hypotheses 1 and 2 look contradictory.  Or does this 

> Only one sweep op is sent during the first attempt.

Mean that during the first attempt, "only one" sweep op is sent? As opposed to multiple ops in a single attempt?

The way I read it at first was that "an op is _only_ sent during the _first_ attempt and not later attempts" which is the direct opposite of number 2.

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 5]
```diff
@@ -770,10 +770,16 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
...
 					eventName: "DeletedAttachmentBlobNotFound",
+					category: alreadyDeleted ? "generic" : "error",
 					blobId,
+					details: { alreadyDeleted },
```
Comment (markfields) [inline packages/runtime/container-runtime/src/blobManager.ts:782]: Redundant with category, but it's probably nice to call it out here. I do worry about log conciseness and volume, but this event is not going to be the deal-breaker 😅

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts [hunk 6]
```diff
@@ -28,34 +41,74 @@ import {
...
+ * Validates that the given blob state is correct in the summary:
+ * - It should be no blob tree in the summary.
+ * - The blob should not be present in the GC state in GC summary tree.
+ * - The blob should be present in the deleted nodes in GC summary tree.
+ */
+function validateBlobStateInSummary(
+	summaryTree: ISummaryTree,
+	blobNodePath: string,
+	expectDelete: boolean = true,
+	expectGCStateHandle: boolean = false,
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts:58]: I'd make these required

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts [hunk 7]
```diff
@@ -1086,4 +1126,246 @@ describeCompat("GC attachment blob sweep tests", "NoCompat", (getTestObjectProvi
...
+		 * 1. Overrides the summarize function of the given container runtime to fail until final summarize attempt.
+		 *
+		 * 2. If "pauseInbound" is true, pauses the inbound queue until the final summarize attempt is completed.
+		 *
+		 * 3. Generates and returns a promise which resolves with ISummarizeEventProps on successful summarization.
+		 */
+		async function overrideSummarizeAndGetCompletionPromise(
+			summarizer: ISummarizer,
+			containerRuntime: ContainerRuntime,
+			pauseInbound: boolean = false,
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts:1143]: Default of true would match typical production behavior, right? Or this is saying keep it paused even in between attempts?  Then never mind.

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts [hunk 8]
```diff
@@ -1086,4 +1126,246 @@ describeCompat("GC attachment blob sweep tests", "NoCompat", (getTestObjectProvi
...
+		) {
+			let latestAttemptProps: ISummarizeEventProps | undefined;
+			const summarizePromiseP = new Promise<ISummarizeEventProps>((resolve) => {
+				const handler = (eventProps: ISummarizeEventProps) => {
+					latestAttemptProps = eventProps;
+					if (eventProps.result !== "failure") {
+						summarizer.off("summarize", handler);
+						resolve(eventProps);
+					} else {
+						assert(
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts:1153]: Not sure where this assert will "go".  I guess to the `emit("summarize")` call.  Will that properly fail your test?  You may want to do `if (...) { reject("Unexpected summarization failure"); }` instead

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts (hunk 7 above)
Comment (markfields) [reply packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts:1143]: Maybe rename it to `blockInboundGcOp` or something?

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts [hunk 9]
```diff
@@ -1086,4 +1126,246 @@ describeCompat("GC attachment blob sweep tests", "NoCompat", (getTestObjectProvi
...
+			const summarizeOverride = async (options: any) => {
+				summarizeFunc = summarizeFunc.bind(containerRuntime);
+				const results = await summarizeFunc(options);
+				if (
+					latestAttemptProps === undefined ||
+					latestAttemptProps.maxAttempts - latestAttemptProps.currentAttempt > 1
+				) {
+					throw new RetriableSummaryError(summarizeErrorMessage, 0.01);
+				}
+				if (pauseInbound) {
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts:1182]: nit - call out it's the last attempt here, something like this:

```suggestion
				// else, this is the last attempt. Resume the queue if needed and return.
				if (pauseInbound) {
```

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts [hunk 10]
```diff
@@ -1086,4 +1126,246 @@ describeCompat("GC attachment blob sweep tests", "NoCompat", (getTestObjectProvi
...
+						defaultMaxAttemptsForSubmitFailures,
+						`The summary should have succeeded at attempt number ${defaultMaxAttemptsForSubmitFailures}`,
+					);
+					assert.equal(
+						gcSweepOpCount,
+						gcOps === "single" ? 1 : props.currentAttempt,
+						`Incorrect number of GC ops`,
+					);
+
+					// The GC sweep will be processed during the multiple summarize attempts and the final summary
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts:1312]: typo


```suggestion
					// The GC Sseep op will be processed during the multiple summarize attempts and the final summary
```

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts [hunk 11]
```diff
@@ -1086,4 +1126,246 @@ describeCompat("GC attachment blob sweep tests", "NoCompat", (getTestObjectProvi
...
+
+		/**
+		 * In these test, summarize fails until the final attempt but GC succeeds in each of the attempts.
+		 * In case of "single" gcOps, in first attempt, GC sends a single sweep op which is processed before the
+		 * next summarize attempt.
+		 * In case of "multiple" gcOps, in every attempt, GC sends a sweep op with the same deleted blob id.
+		 * It validates that in these scenario, the blob is correctly deleted and nothing unexpected happens.
+		 */
+		for (const gcOps of ["single", "multiple"]) {
+			itExpects(
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts:1199]: This test is really cool, nice job

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts [hunk 12]
```diff
@@ -29,48 +31,110 @@ import {
 } from "@fluid-private/test-version-utils";
 import { delay } from "@fluidframework/core-utils";
 import { IContainer, LoaderHeader } from "@fluidframework/container-definitions";
-import { IErrorBase, IFluidHandle } from "@fluidframework/core-interfaces";
-import { getGCDeletedStateFromSummary, getGCStateFromSummary } from "./gcTestSummaryUtils.js";
+import { IErrorBase } from "@fluidframework/core-interfaces";
+// eslint-disable-next-line import/no-internal-modules
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:35]: What's up with these lint disables?

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts [hunk 13]
```diff
@@ -29,48 +31,110 @@ import {
...
+					},
+				},
+				gcOptions,
+			},
+			loaderProps: { configProvider: mockConfigProvider(settings) },
+		};
+	});
+
+	afterEach(() => {
+		settings = {};
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:137]: Don't do it here, but mentioning it since we've been discussing. Here's some code will clear the object without creating a new one:

`Object.keys(x).forEach((k) => delete x[k])`

Maybe there's a better way, it's what I came up with.

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts [hunk 14]
```diff
@@ -157,7 +200,14 @@ describeCompat("GC data store sweep tests", "NoCompat", (getTestObjectProvider)
...
 		});
 		const summarizerDataObject = response.value as ITestDataObject;
-		await sendOpToUpdateSummaryTimestampToNow(summarizer2);
+		await delay(sweepTimeoutMs + 10);
+
+		// Send an op to update the timestamp that the summarizer client uses for GC to a current one.
+		defaultDataObject._root.set("update", "timestamp");
+		await provider.ensureSynchronized();
+
+		// Close the containers as these containers would be closed by session expiry before sweep ready ever occurs
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:209]: nit: update wording (containers shouldn't be plural anymore)

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts [hunk 15]
```diff
@@ -328,8 +375,7 @@ describeCompat("GC data store sweep tests", "NoCompat", (getTestObjectProvider)
 					summarizingContainer,
 					summarizer,
 					summaryVersion: unreferencedSummaryVersion,
-				} = await summarizationWithUnreferencedDataStoreAfterTime(sweepTimeoutMs);
-				await sendOpToUpdateSummaryTimestampToNow(summarizer);
+				} = await summarizationWithUnreferencedDataStoreAfterTime();
 				const sendingContainer = await loadContainer(unreferencedSummaryVersion);
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:379]: I'm not thinking too critically about removing this "send op" call, assuming the tests provide coverage for themselves, essentially.

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts [hunk 16]
```diff
@@ -446,61 +491,17 @@ describeCompat("GC data store sweep tests", "NoCompat", (getTestObjectProvider)
...
 			const { unreferencedId, summarizer } =
-				await summarizationWithUnreferencedDataStoreAfterTime(sweepTimeoutMs);
+				await summarizationWithUnreferencedDataStoreAfterTime();
 			const sweepReadyDataStoreNodePath = `/${unreferencedId}`;
-			await sendOpToUpdateSummaryTimestampToNow(summarizer);
 
 			// Summarize. In this summary, the gc op will be sent with the deleted data store id. The data store
 			// will be removed in the subsequent summary.
-			await summarizeNow(summarizer);
+			await summarize(summarizer);
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:501]: Maybe rename to `syncAndSummarize` or something?  In the past I've gotten tripped up on the lack of `ensureSynchronized` before when reading tests in this file I think

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts [hunk 17]
```diff
@@ -541,23 +541,245 @@ describeCompat("GC data store sweep tests", "NoCompat", (getTestObjectProvider)
...
+
+		/**
+		 * This function does the following:
+		 * 1. Overrides the summarize function of the given container runtime to fail until final summarize attempt.
+		 *
+		 * 2. If "pauseInbound" is true, pauses the inbound queue until the final summarize attempt is completed.
+		 *
+		 * 3. Generates and returns a promise which resolves with ISummarizeEventProps on successful summarization.
+		 */
+		async function overrideSummarizeAndGetCompletionPromise(
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:573]: If this is identical to the on in the Blobs test file, move to the utils file?

File: packages/runtime/container-runtime/src/summary/runningSummarizer.ts [hunk 18]
```diff
@@ -842,12 +867,20 @@ export class RunningSummarizer extends TypedEventEmitter<ISummarizerEvents> impl
...
 
 		const { reason, ...summarizeOptions } = options;
-		const result = this.trySummarizeOnce(
-			{ summarizeReason: `onDemand/${reason}` },
-			summarizeOptions,
-			resultsBuilder,
-		);
-		return result;
+		if (options.retryOnFailure === true) {
+			this.summarizeOnDemandWithRetries(`onDemand;${reason}`, resultsBuilder).catch(
```
Comment (markfields) [inline packages/runtime/container-runtime/src/summary/runningSummarizer.ts:871]: This uses `;` but the else case uses `/`

File: packages/runtime/container-runtime/src/summary/runningSummarizer.ts [hunk 19]
```diff
@@ -842,12 +867,20 @@ export class RunningSummarizer extends TypedEventEmitter<ISummarizerEvents> impl
...
+				(error) => {
+					resultsBuilder.fail("summarize failed", error);
```
Comment (markfields) [inline packages/runtime/container-runtime/src/summary/runningSummarizer.ts:873]: This will never do anything, `resultsBuilder.build()` will have already been called and returned.

Comment (markfields) [reply packages/runtime/container-runtime/src/summary/runningSummarizer.ts:873]: Ok never mind, I missed two things:

1. `resultsBuilder` is passed in, so caller continues to monitor it (awaits the promises)
2. `resultsBuilder.build()` just packs up the promises into a convenient object. I assumed it tore-down the builder and gave some independent result.

So a caller can await the promises returned by this function, and whenever `summarizeOnDemandWithRetries` fails those promises will resolve.

Seeing that `resultsBuilder` is passed into `summarizeOnDemandWithRetries`, it's a little surpising you need this catch block to do anything.  I'll take a closer look there.

Comment (markfields) [reply packages/runtime/container-runtime/src/summary/runningSummarizer.ts:873]: Ok. As we've discussed before it would be sweet to refactor this stuff around these 3 promises.  It is always hard to follow.  I think this approach makes sense though.

Do your tests cover this path?  Maybe put a breakpoint on this line and make sure it's hit and is proven to work as expected.

**Review by markfields — APPROVED** [review]

**Review by markfields — APPROVED** [review]

File: packages/runtime/container-runtime/src/summary/runningSummarizer.ts (hunk 19 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/summary/runningSummarizer.ts:873]: Yeah, it wasn't straightforward to make it work and I don't like the solution. But this is the best I came up with. Good point about testing it works as expected. I will add that.

**Comment by agarwal-navin** [comment]
> @agarwal-navin - Looking at the description, test hypotheses 1 and 2 look contradictory. Or does this
> 
> > Only one sweep op is sent during the first attempt.
> 
> Mean that during the first attempt, "only one" sweep op is sent? As opposed to multiple ops in a single attempt?
> 
> The way I read it at first was that "an op is _only_ sent during the _first_ attempt and not later attempts" which is the direct opposite of number 2.

Updated the description to make it clearer.

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts (hunk 6 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts]: done.

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts (hunk 7 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts]: renamed

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts (hunk 8 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcSweepAttachmentBlobs.spec.ts]: Oh man. I have done this before too! Fixed it, thanks!

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts (hunk 12 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:40]: Fixed it as per the new conventions.

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts (hunk 15 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:385]: It was not required. The summarizationWithUnreferencedDataStoreAfterTime does that already before returning. Calling it again was just redundant.

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts (hunk 16 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts]: Renamed to `ensureSynchronizedAndSummarize`

File: packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts (hunk 17 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcSweepDataStores.spec.ts:586]: I think we should merge these two test files into one. I am going to leave it as-is for now.

File: packages/runtime/container-runtime/src/summary/runningSummarizer.ts (hunk 18 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/summary/runningSummarizer.ts:871]: Yeah, I know. I tried changing it to `onDemand;` in the past for the other one but it started failing PR pipelines for some reason. I could not figure out why so I reversed that change. See https://github.com/microsoft/FluidFramework/pull/16886 for details.
